### ISSUE TITLE
[BUG] `term_vector` does not work on `search_as_you_type` fields

### ISSUE THREAD

**Issue by haversnail** [opensearch-project/OpenSearch#1901]
**Describe the bug**
TL;DR: same issue reported [here](https://github.com/elastic/elasticsearch/issues/66407).

Per ElasticSearch's 7.10 [documentation](https://www.elastic.co/guide/en/elasticsearch/reference/7.10/search-as-you-type.html#general-params) (which AFAICT should be compatible with OpenSearch 1.x), setting `term_vector` on a `search_as_you_type` field should only affect the shingle subfields, but not the index prefix subfield. However, when attempting to index documents with this mapping, an `illegal_argument_exception` is thrown.

**To Reproduce**
1. Create a new index with the following mapping:
```jsonc
{
  // ...
  "my_field":
    "type": "search_as_you_type",
    "term_vector": "with_positions_offsets"
  // ...
}
```
2. Attempt to index a document containing a `my_field` field
3. Observe the following error:
```json
{
  "type": "illegal_argument_exception",
  "reason": "cannot index term vector offsets when term vectors are not indexed (field=\"my_field._index_prefix\")"
}
```

**Expected behavior**
Documents should be indexed without issue. Only the `._2gram` and `._3gram` shingle subfields should have the `term_vector` set.

**Plugins**
N/A

**Screenshots**
N/A

**Host/Environment**
 - Version: 1.2.3 (opensearchproject/opensearch:latest)

**Comment by haversnail** [comment]
I noticed there was a PR that claims to resolve the issue, however I just tried this with the latest version of OpenSearch and am still encountering the same issue. ☹️  
```
type: 'illegal_argument_exception',
reason: 'cannot index term vector offsets when term vectors are not indexed (field="name._index_prefix")'
```

**Comment by dblock** [comment]
@vpehkone ^

**Comment by vpehkone** [comment]
I tested with below OpenSearch docker versions and single node cluster. Didn't get errors after 2.0.0 version.

2.7.0: no errors
2.6.0: no errors
2.5.0: no errors
2.4.0: no errors
2.3.0: no errors
2.2.0: no errors
2.1.0: no errors
2.0.0: got the error

Test commands:
```
curl -XPUT -H "Content-type: application/json" http://localhost:9200/myindex?pretty -d '{ "mappings": { "properties": { "myfield": { "type": "search_as_you_type", "term_vector": "with_positions_offsets" }}}}'
```
```
curl -XPOST -H "Content-type: application/json" http://localhost:9200/myindex/_doc/1?pretty -d '{"myfield":"XXXX"}'
```
Responses if no error:
```
{
  "acknowledged" : true,
  "shards_acknowledged" : true,
  "index" : "myindex"
}
```
```
{
  "_index" : "myindex",
  "_id" : "1",
  "_version" : 1,
  "result" : "created",
  "_shards" : {
    "total" : 2,
    "successful" : 1,
    "failed" : 0
  },
  "_seq_no" : 0,
  "_primary_term" : 1
}
```
Responses if error:
```
{
  "acknowledged" : true,
  "shards_acknowledged" : true,
  "index" : "myindex"
}
```
```
{
  "error" : {
    "root_cause" : [
      {
        "type" : "illegal_argument_exception",
        "reason" : "cannot index term vector offsets when term vectors are not indexed (field=\"myfield._index_prefix\")"
      }
    ],
    "type" : "illegal_argument_exception",
    "reason" : "cannot index term vector offsets when term vectors are not indexed (field=\"myfield._index_prefix\")"
  },
  "status" : 400
}
```

**Comment by dblock** [comment]
@haversnail which version have you tried? any different from what @vpehkone is doing?

**Comment by sejli** [comment]
@YashAnetz will be trying this one out for OSCI. Great starting point would be the YAML test that dblock@ mentioned

**Comment by msfroh** [comment]
I could be mistaken, but I think a fix similar to the one from https://github.com/opensearch-project/OpenSearch/pull/12048 might help here.

Specifically, I think the issue is that it's trying to create a query on one of these "internal" fields. It should probably add a check to see if the field name contains a dot (`.`), and if so, is the character following the last dot an underscore(`_`). If so, it's an internal field and should be skipped.

**Comment by getsaurabh02** [comment]
We discussed this in the Search community meeting today. Eric Pugh volunteered to take a stab it.

**Comment by epugh** [comment]
I've got a build running on my Mac now, and am more comfortable pickign this one up!

**Comment by diogopontual** [comment]
I was looking for a good first issue to work on and ended up here.
I followed the steps described by @haversnail  using the current main branch, and it worked like a charm.
I saw @msfroh ’s comment about the dots in the field name — happy to help if needed!

**Comment by ThyTran1402** [comment]
Hi @sandeshkr419 @epugh 

I was able to find the root causes of these issues.
+ The core bug is on the backend for search_as_you_type is a compound field type. When you define one, OpenSearch automatically creates three internal subfields: my_field._2gram, my_field._3gram, my_field._index_prefix. OpenSearch propagates term_vector: "with_positions_offsets" from the parent field to all subfields, including _index_prefix. But _index_prefix uses an internal prefix index structure that is fundamentally incompatible with term vector storage (offsets/positions). The engine throws illegal_argument_exception at index time.
+ Another bug is on OpenSearch Dashboards are search_as_you_type is missing from OPENSEARCH_FIELD_TYPES enum, Console autocomplete doesn't list it, term_vector is treated as universally applicable, _index_prefix subfield is blindly handled as a multi-field

The ways to fix it: 
+ On OpenSearch core: don't propagate term_vector to PrefixFieldMapper — only apply it to ShingleFieldMapper subfields
+ Dashboards types: add SEARCH_AS_YOU_TYPE = 'search_as_you_type' to OPENSEARCH_FIELD_TYPES   
+ Dashboards console: add 'search_as_you_type' to the type autocomplete list 
+  Dashboards field caps: consider filtering out internal _index_prefix subfields the same way object/nested parents are already filtered
- I will need to fix no document-indexing test (regression risk) and no yaml rest integration test
Let me know if the fixes work, I will start working on it.
Thank you!

**Comment by ThyTran1402** [comment]
Can someone please review it?  @epugh @dblock 

Thank you.
